sklearn/utils/_pprint.py

```diff
@@ -89,18 +89,26 @@ def _changed_params(estimator):
     estimator with non-default values."""
 
     params = estimator.get_params(deep=False)
-    filtered_params = {}
     init_func = getattr(estimator.__init__, 'deprecated_original',
                         estimator.__init__)
-    init_params = signature(init_func).parameters
+    init_params = inspect.signature(init_func).parameters
     init_params = {name: param.default for name, param in init_params.items()}
 
-    for k, v in params.items():
-        if (k not in init_params or (  # happens if k is part of a **kwargs
-                repr(v) != repr(init_params[k]) and
-                not (is_scalar_nan(init_params[k]) and is_scalar_nan(v)))):
-            filtered_params[k] = v
-    return filtered_params
+    def has_changed(k, v):
+        if k not in init_params:  # happens if k is part of a **kwargs
+            return True
+        if init_params[k] == inspect._empty:  # k has no default value
+            return True
+        # avoid `repr` on nested estimators, it causes nonlinear complexity
```

Comment:
```suggestion
        # try to avoid calling repr on nested estimators
```

Reply:
I added the test and verified it fails as expected when any of the two fixes is commented out.